Feature/audio player

riban-bw opened the pull request:
Adds bespoke audio player as a Zynthian engine. This allows playback of wav, ogg, etc. files through effects chains to a mixer strip. Files are presented as presets and folders as banks.

riban-bw committed:
Wireframe for audio player. Loads files with info and presents stereo jack source but does not play yet.

riban-bw committed:
Adds function to get duration of an audio file without loading it into the player

riban-bw committed:
Fix get file duration when file does not exist

riban-bw committed:
Adds playback to audioplayer (wip)

riban-bw committed:
Fixes build audio player build script

riban-bw committed:
Adds samplerate conversion to audio player

riban-bw committed:
Merge branch 'testing' into feature/audio_player

riban-bw committed:
Audio Player Library:
Implemented loop mode.
Fix filename handling.
Implement setPosition. (Now uses seconds rather than frames as argument.)
Add get_filename.
Fix and optimise playback.
Use medium quality samplerate conversion to allow RPi3 to cope.
Use stream samplerate conversion.
Add get_info.

riban-bw committed:
Audio Player:
Added copyright notices to source.
Reduce buffer size to allow more promt seek.
Use floats rather than doubles for durations. (We only really care about 0.1s accuracy most of the time.)
Don't stop playback when seeking.
Use fastest samplerate conversion to improve seek times, especially on RPi3.
Start double-buffer population from active buffer (rather than buffer 0) - probably of no benefit.
Add getPosition function.

riban-bw committed:
Merge branch 'testing' into feature/audio_player

riban-bw committed:
Merge branch 'testing' into feature/audio_player

riban-bw committed:
Audio player:
Fix compile error.
Some optimisation - wip.

riban-bw committed:
Adds ability to set audioplayer SRC quality

riban-bw committed:
Merge branch 'feature/audio_player' of github.com:zynthian/zynthian-ui into feature/audio_player

riban-bw committed:
Reverts OSC path for mixer fader.Reverts OSC path for mixer fader.

riban-bw committed:
Add volume control to audio player library

riban-bw committed:
Merge testing into audioplayer branch

riban-bw committed:
Implementing zynaudiplayer

Add zynaudioplayer to Audio Generators.
Set as single MIDI channel engine.
Convert zyanaudioplayer from c++ to c.
Fix hang / crash on exit.
Add MIDI CC control of zynaudioplayer.

riban-bw commented:
This will resolve zynthian/zynthian-issue-tracking#645. There are a few things yet to do that I may need some help with, e.g. setting the controls from the library / engine.

riban-bw committed:
Added missing audioplayer engine file.

riban-bw committed:
Adds ability to set position of audio player

riban-bw committed:
Add ability to rename and delete audio files

jofemodo committed:
Merge branch 'testing' into feature/audio_player

riban-bw committed:
Adds mulitchannel audio recorder to chain context menus

riban-bw committed:
Merge branch 'feature/audio_player' of github.com:zynthian/zynthian-ui into feature/audio_player

riban-bw committed:
Change zynaudioplayer to use ring buffer instead of double buffer.
Note: Still issues with samplerate conversion.
Remove default priming of main output to recorder.
Will use main output if none other primed.

riban-bw committed:
Zynaudioplayer set samplerate from jack at init

riban-bw committed:
Zynaudioplayer fixes.

Fixed playback glitches.
Fix oversized buffers.

riban-bw committed:
Adds playback mode to audio player

riban-bw committed:
zynaudioplayer fixes.

Fixed looping.
Soft mute when stopped.

riban-bw committed:
Recue audioplayer at end of playback

riban-bw committed:
Use symbol to show audio recorder status in layer menu.
Remove audio recorder from main menu.

riban-bw committed:
Allow definition of audio player jack client name

riban-bw committed:
Merge branch 'feature/audio_player' of github.com:zynthian/zynthian-ui into feature/audio_player

riban-bw committed:
Allow multiple instances of audio player

riban-bw committed:
Use jack_ringbuffer in audio player.
Minimum soundfile reduced to 10 frames (maybe can reduce to 0).
Fix bug starting playback of file causing crash.
Simplified end of file detection.
Only recue to zero if file played to end.

riban-bw committed:
Remove audio player minimum file size.
Limit set_position to length of file.

riban-bw committed:
Fixes to audio player

Set loop during playaback would reset to start - fixed.
Use "best quality" samplerate conversion as default.

riban-bw committed:
Revert audio player default SRC quality to "fastest"

riban-bw committed:
Initialise filename in audio player

riban-bw committed:
Add ability to change audio player buffer size and count

riban-bw committed:
Add OSC notifications to audio player.
Fix position wrap on loop for samplerate converted files.

riban-bw committed:
Audio player enhancements.

Add waveform display with playhead indicator.
Fix snapshot saving.
Add audio recorder channel prime to snapshot.
Move OSC notification to own thread in mixer library.
Increase OSC heartbeat timeout from 12s to 120s.

riban-bw committed:
Add tinyosc source files

riban-bw committed:
Adds audio player widget

jofemodo committed:
Remove obsolete jackpeak directory.

riban-bw committed:
Audio Player enhancements

Changed colours of waveform display.
Added OSC control (not using within zynthian controllers yet).
Changed "open" to "load" and "close_file" to "unload".

riban-bw committed:
Merge branch 'feature/audio_player' of github.com:zynthian/zynthian-ui into feature/audio_player

riban-bw committed:
Minor tweaks to audio player.

Renames source files.
Fixes debug output references (now all refer to libzynaudioplayer.
Added OSC implementation doc.

riban-bw committed:
Adds float option to zynthian_controller.
Add OSC server to audio player (not necessary but nice to have).
Hide waveform controls for bad audio files.
Control audio player directly rather than via MIDI or OSC.
Fix some bugs in zynthian_gui_controller.
libzynaudioplayer now waits for file to open before returning from load().
Add get_src_quality() to libzynaudioplayer.

riban-bw committed:
Remove LV2 dependancy from default widget.
Show file duration in waveform display of audio player.

riban-bw committed:
Adds note on/off playback of audio player

riban-bw committed:
Improve note-on playback of audio player

riban-bw committed:
Adds pitch to MIDI note playback of audio player.

C4 (MIDI 60 / middle C) is no pitch change.
Has issues playing back slower than unity (below C4).
Playback speed is changed - maybe better to stretch sound but not yet....

riban-bw committed:
Do not send controller values when showing control screen. (Was causing parameters to change unexpectedly.)
Add access to debug / config elements in audio player (temporary for current dev cycle).
Fixed play cursor in audio player when playing back at different speed.
Started to add pitch bend to audio player but that is unlikely to progress.

jofemodo committed:
Fix font size configuration.

jofemodo committed:
+ Change waveform color.
+ Duration text: move to right and reduce font size.

riban-bw committed:
Audio Player enhancements

Fix track selection of multi-track files.
Can now individually select track to play to each output.
Show file samplerate and axis labels on waveform.
Rename 'playback_track' functions to 'track'.

riban-bw committed:
Disable OSC support in Audio Player by default

riban-bw committed:
Fix optional OSC support for audio player

riban-bw committed:
Fixes and enhancements to audio player.

Show one level of file structure (directories) in Audio Player.
Change 'play' control to 'transport' - will affect snapshots.
Implement event driven notification of changes.
Remove user control of buffer configuration.
Fix rename file (was missing '.' before ext).
Allow delete/rename of files on USB.
Fix crash on opening empty rf bad file. Now show indication.
Fix inconsitent behaviour / crash if trying to access closed file.
Disable MIDI control by default.

riban-bw committed:
Changes to implement model-view-controller system.

This is required to support update of view from changes in engines, e.g. audio player transport.
Moved master data to zynthian_controller class.
Controllers now send data to model.
Model updates views (events) or views get data from model (poll).
Consolidated many user interfaces, e.g. keyboard, mouse, touch, encoders, etc. for unified interface.
Fixes several UI issues like list scrolling with mouse wheel, selection from list with touch, etc.
Know issues:
- Initial display of arc is sometimes too short
- Encoder acceleration is always enabled
- Logarithmic scales not validated (need to know where used to allow testing)

riban-bw committed:
Fixes / improves scaling of float controls

riban-bw committed:
Fix MIDI learn

riban-bw committed:
Remove control view widget thread. Now updates from controller refresh.
Previously new threads were created for each widget which is overkill.

riban-bw committed:
Optimise DPM update (reduce mixer view CPU usage per channel)

riban-bw committed:
Small optimisation of DPM.
Change mono DMP colour to better match stereo colour.

riban-bw committed:
Adds OSC control of mixer mono state

jofemodo commented on zyngui/zynthian_gui_control.py:
```diff
@@ -330,21 +330,79 @@ def back_action(self):
+			except:
+				pass
+		else:
+			super().select_up()
```
No my friend!!! You can't change the behaviour of things you don't understand like this!! Are you missing we have a new prototype called Z2? I will revert this change. Please, don't change things like this. I will send you a Z2 ASAP so you can test with it. Until then, please, think wise. There are lots of improvements & adjustments specifically done for the Z2.

jofemodo replied:
Also, this is very VX-centric. We are moving to something more flexible. The classic "4 rotaries + 4 switches" still have to work, but it shouldn't be hardcoded anymore. I mean, we shouldn't hardcode switch/knobs actions like this.  Not in new code. In fact, we should refact old code to remove this.

riban-bw replied:
I am aware that these changes are likely to break stuff and have been methodically working through the modules to fix errors. Of course it is challenging to test the Z2 stuff but I had intended to look at its behaviour but tried to make the initial changes hardware agnostic.

VX-centric? Not sure what that means.

riban-bw committed:
Fix contrlloler always sending discrete values.
Audio player transport play now >63 (was >0).
Remove unused function zynthian_gui_control:set_controller_value_by_index.
Update selector value.
Fix rentrant issue with selector.

jofemodo committed:
Restore old functionality & behaviour, so Z2 can work flawlessly.

jofemodo committed:
Merge branch 'feature/audio_player' of github.com:zynthian/zynthian-ui into feature/audio_player

jofemodo committed:
Fix typo.

riban-bw committed:
Adds mouse wheel on waveform control of position in audio player

jofemodo commented:
Controller refact "side effects" (part I)

**setbfree**
    + "Rotary Speed" controller doesn't work like expected. Can't change to rotary fast and "off" is "slow", etc.
    + "DrawBars" are inverted, When you move up, they go down. When you move down, you go up.

jofemodo commented:
BTW, i just noted that AudioPlay CUIAs doesn't work anymore. This also breaks quite a bit the Z2 behaviour.
```
		#TODO: Handle CUIA audio playback
		#elif cuia == "START_AUDIO_PLAY":
		#elif cuia == "STOP_AUDIO_PLAY":
		#elif cuia == "TOGGLE_AUDIO_PLAY":
```
It's not a big problem, but i would need to know if you already have plans for restoring this and how do you intend to do it.

jofemodo committed:
+ Add back_action function to StepSequencer, so it can work OK with Z2.
+ Make Z2 cursors to work with StepSequencer (Zynpad & Pattern Editor)

jofemodo committed:
Add new ARROW_XXX CUIAs, keeping the legacy 4 x rotary UP/DOWN actions.

riban-bw commented:
Playback is per instrument and can be triggered with CC. It may be part of the API to trigger such actions similar to CUIA. The old CUIA for audio player was not obvious which audio it would play. The new player may have many instances, each with own CC to control via MIDI learn for example. I suggest we don't implement CUIA for audio player.

I am methodically working through functional tests. I have not yet reached setBfree (beyond checking it loads and plays).

I didn't realise / had forgotten there was a PR for this development branch hence have been working heavily on it.

jofemodo committed:
Add new ARROW_XXX CUIAs to zynconf and midi key list.

jofemodo committed:
Fix hmode for toggle_screen CUIAs.

jofemodo commented:
I think you need a Z2 urgently ...  ;-)
I hope to send you one very soon, but they are difficult to get, mate!!

jofemodo commented:
**Selector Separators:**
It seems to be some problem with "skip_separators" on zynthian_gui_selector. Try going slowly throw a separator, it start to jump crazy up & down. This is due to the recent refact for sure ;-)

jofemodo commented:
**Navigating Controller Screens** with the selector knob doesn't preview the controllers pages like before.

riban-bw committed:
Fix controller sometimes not updating.
Add phase reverse to mixer.
Fix mouse wheel on audio player waveform.
Fix error destroying audio player library.
Catch possible segfault in mixer for channels > max.

riban-bw committed:
Merge branch 'feature/audio_player' of github.com:zynthian/zynthian-ui into feature/audio_player

riban-bw committed:
Fixes audio player wrong audio output selection after loading file

jofemodo committed:
Fix controller-page navigation preview.

jofemodo committed:
Merge branch 'feature/audio_player' of github.com:zynthian/zynthian-ui into feature/audio_player

jofemodo commented:
> It is not possible to select mixer chain with QWERTY keyboard. Previously this was done with the up/down arrow keys emulating the select encoder.

You should use the new ARROW_LEFT/RIGHT CUIAs.

jofemodo commented:
Navigating Controller pages & skip_separators seems to be solved with my latest commit.

jofemodo committed:
Fix "rotary speed" controller.

riban-bw committed:
Add global dynamically instantiated audio player.
Enable CUIA for audio player.
Adds function to route audio to aux input of mixer.
Removes obsolete (and broken) zynthian_engine::get_cmdlist function.
Do not select control page when scrolling listbox with mouse wheel (expected UX).
Update controls whilst selecting page in 'select' mode.
Don't call callback in audioplayer library if player removed.

riban-bw committed:
Merge branch 'feature/audio_player' of github.com:zynthian/zynthian-ui into feature/audio_player

riban-bw committed:
Fix divide by zero error when controller range calculated as zero.
Fix play status indication of global audio player.

riban-bw committed:
Fixes key range / transpose - requires update to zyncoder git push ff3c222
Adds ability to toggle MIDI learn in key range / transpose view.
Send all notes off CC before changing transpose to avoid stuck notes.
Fixed tempo control in MIDI Recorder view.
Fixed hanging notes when stopping MIDI player.
Fixed global player failing if expected file is deleted during a session.
Fixes XY control view.
Show appropriate significant digits of float values on controls.
Remove obsolete functions.

riban-bw committed:
Adds record control to audio player.
Removes quality from audio player.
Moves gain control to second page.

riban-bw committed:
Move gain control in audio player.
Fix audio player bugs.
Permit None as controller to allow non-consecutive layout.
Fix exception if controller not assigned.
Fix mouse wheel scrolling on waveform when fewer than 4 controls.
Fix too may loops being played when looping disabled.
Added loop points to audio player library (not yet exposed).
Add ability to define playback time delta for notifications. (May reduce chatter.)

riban-bw committed:
Adds general purpose status indicator for layers.
Implemented audio player playback status indication in mixer using new status indicators.
Added loop points to audio player ***wip - changing loop end time can trigger seg fault and many other issues to resolve***
Hide multichannel recorder options by default. Can be enabled by settting env var ZYNTHIAN_UI_MUKTICHANNEL_REC.
Fixed bug plotting controllers without control.
Remove audio record control from chain options menu.
Change name of main menu items removing "FX"
Moves audio player to Audio Chain list

riban-bw committed:
Fixes seg fault in audio player

jofemodo committed:
Refactorize for relative zynpots (partial 1).

jofemodo committed:
Fix typo.

jofemodo committed:
Refactorize for zynpot's relative behaviour (partial 2).

jofemodo committed:
+ Refact mixer for working with relative zynpots (partial 3).
+ Increase control thread frequency to 100 while keeping the GUI refresh frequency at 25.

jofemodo committed:
Fixes the last refactorization.

jofemodo committed:
Replace "text_only" variable by "selector_counter".

jofemodo committed:
Fix selector GUI controllers.

jofemodo committed:
+ Fix key_range midi learning for working with Z2 by implementing a new GUI class function: toggle_midi_learning().

jofemodo committed:
Refact stepsequencer for working with relative zynpots.

jofemodo committed:
Refact GUI keyboard for working with relative zynpots.

riban-bw committed:
Audio player fixes

Fixes seg fault when setting loop end on short audio files.
Fixes indication of play position when playing faster / slower.
Fixes playback at slower than 1x speed. This has removed SRC so may alias.

riban-bw committed:
Merge branch 'feature/audio_player' of github.com:zynthian/zynthian-ui into feature/audio_player

riban-bw committed:
Limit position of audio player to loop points

riban-bw committed:
Audio Player enhancements and fixes.

Move play position to within loop markers if loop enabled.
Catch end of buffer during playback.
Only loop to start if playing.
Disable loop points (until resolve issues).

riban-bw commented:
I think this is probably good enough to merge. Issues with encoder control, etc. should be tracked and resolved under separate tickets / dev branches. I will report such issues in github issue tracker.

riban-bw committed:
Fix logarithmic control values.
Colour value and title when setting XY mode so can see which control is selected when value is at minimum.
Add whitespace and update copyright year for some code.

jofemodo committed:
+ Update FontAwesome & install ForkAwesome TTF font.
+ Replace solo & mute button's text on mixer + Improve colors

riban-bw committed:
Fixes unexpected navigation from select menus

riban-bw committed:
Fixes unexpected navigation from select menus.

jofemodo committed:
Increase contrast on mixer's solo/mute buttons.

riban-bw committed:
Fix exception when opening invalid file in audio player.
Scale waveform to fit display in audio player.
Show context sensitive labels on buttonbar.
Stop learning when exiting control view.

riban-bw committed:
Reduce brightness of solo/mute buttons to improve contrast and ledgibility

jofemodo committed:
Minor color adjustment on mixer's solo & mute buttons.

jofemodo committed:
Refactorize & Implement an event model for zynpots: No more polling the zynpots!

jofemodo committed:
Revert last changes in color-scheme for "unhighlighted" GUI controllers.
